### PR TITLE
[ILM] add docs explaining high-level concepts

### LINKED ISSUE #34936: ILM Docs - Add Documentation describing Concepts
It is important to explain the building blocks of ILM: policies, actions, phases

Policies
Phases - hot, warm, cold, delete
minimum_age
Actions - explain all the existing actions in a descriptive way (not death by parameters) (link to reference docs). Explain safe vs. unsafe actions

### PR DESCRIPTION
**Pull request by talevy**
these docs add definitions explaining the different
phases and actions available in ILM

Closes #34936.

### CODE REVIEW

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+[source,js]
+--------------------------------------------------
+{
+  "allocate" : {
+    "number_of_replicas" : 2
+  }
+}
+--------------------------------------------------
+// NOTCONSOLE
```
Comment (talevy) [inline docs/reference/ilm/policy-definitions.asciidoc]: Would we prefer real CONSOLE examples here? I hesitated at first due to the extra noise, but I also see the value of running these snippets and giving further context about which phases these actions are defined in. What do you think?

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+--------------------------------------------------
+{
+  "shrink" : {
+    "number_of_shards": 1
+  }
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+== Phase Definitions
```
Comment (talevy) [inline docs/reference/ilm/policy-definitions.asciidoc]: I chose to give a high-level overview of which actions are allowed in which phases and deliberately did not mention this information within each action description. I see value in also mentioning this alongside the action description, but didn't know if this would be too unmaintainable moving forward if these change over time.

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  - <<ilm-rollover-action,Rollover>>
+* Warm
+  - <<ilm-allocate-action,Allocate>>
+  - <<ilm-readonly-action,Read-Only>>
+  - <<ilm-forcemerge-action,Force Merge>>
+  - <<ilm-shrink-action,Shrink>>
+* Cold
+  - <<ilm-allocate-action,Allocate>>
+* Delete
+  - <<ilm-delete-action,Delete>>
```
Comment (talevy) [inline docs/reference/ilm/policy-definitions.asciidoc]: Is here a good place to go into details about the order of execution and conditions on 
execution?

for example:

```
The order in which phases are defined in a policy definition
can be arbitrary, but the order in which they are executed is
pre-determined by {ILM}. The order of execution is: Hot, Warm, Cold, Delete.

Phases can optionally set a `min_age` condition, which determines what
age an index must be before entering the associated phase.
```

**Review by dakrone — CHANGES_REQUESTED** [review]
I left some comments for a first pass, thanks for writing this up Tal

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 4]
```diff
@@ -0,0 +1,293 @@
+[role="xpack"]
+[testenv="basic"]
+[[index-lifecycle-management]]
+= Policy Phases and Actions
+
+There are four stages in the index lifecycle:
+
+* Hot--the index is actively being updated and queried.
+* Warm--the index is no longer being updated, but is still being queried.
```
[the remark this answers is not in the document]
Comment (colings86) [reply docs/reference/ilm/policy-definitions.asciidoc]: I think we should detail the high level aim of each phase here and that is the warm phase is intended for indexes you are done writing to but are still querying. If users want to use it outside of this then thats fine but I have a hard time explaining what the phases are without describing the common workflow

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+* Cold--the index is no longer being updated and is seldom queried. The
+information still needs to be searchable, but it's okay if those queries are
+slower.
+* Delete--the index is no longer needed and can safely be deleted.
+
+Each of these stages is called a "phase". A policy does not need to configure
+each phase for an index. For example, one index may only go from the hot phase to
+the delete phase, while another may run through all phases. Depending on which
```
Comment (colings86) [inline docs/reference/ilm/policy-definitions.asciidoc]: I think this is a bit confusing as is because it might seem like the same policy will be executed differently on different indexes. Maybe instead we should say something like the following?
```
For example, one policy may define only the hot phase and
the delete phase, while another may define all four phases.
```

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+phase one is in, different actions are allowed to be taken.
+
+== Actions
+
+Before discussing which actions are allowed in which phase, it is important
```
[the remark this answers is not in the document]
Comment (colings86) [reply docs/reference/ilm/policy-definitions.asciidoc]: +1

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+to understand all the actions as they are.
+
+[[ilm-allocate-action]]
+=== Allocate
+
+The Allocate action allows you to specify which nodes are allowed to host the
+shards of the index and set the number of replicas.
+Behind the scenes, it is modifying the index settings
+for shard filtering and/or replica counts. When updating number of replicas,
```
Comment (colings86) [inline docs/reference/ilm/policy-definitions.asciidoc]: `When updating number` -> `When updating the number`

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+configuring allocation rules is optional. When configuring allocation rules,
+setting number of replicas is optional. Although this action can be treated as
+two separate index settings updates, both can be configured at once.
```
Comment (colings86) [inline docs/reference/ilm/policy-definitions.asciidoc]: I don't follow this last sentence because we perform this action in one settings update?

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+[[ilm-allocate-options]]
+.Allocate Options
+[options="header"]
+|======
+| Name                 | Required  | Default     | Description
+| `number_of_replicas` | no        | -           | The number of replicas to assign to the index
+| `include`            | no        | -           | assigns an index to nodes having at least _one_ of the values
+| `exclude`            | no        | -           | assigns an index to nodes having _none_ of the values
+| `require`            | no        | -           | assigns an index to nodes having _all_ of the values
```
Comment (colings86) [inline docs/reference/ilm/policy-definitions.asciidoc]: `values` -> `attributes`

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 1 above)
Comment (colings86) [reply docs/reference/ilm/policy-definitions.asciidoc]: I agree with @dakrone. Also having NOTCONSOLE examples means we may miss places to update the documentation if/when things change in the future

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+{
+  "delete" : { }
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+[[ilm-forcemerge-action]]
+=== Force Merge
+
+WARNING: This will make the index read-only (see: <<dynamic-index-settings,index.blocks.write>>)
```
[the remark this answers is not in the document]
Comment (colings86) [reply docs/reference/ilm/policy-definitions.asciidoc]: Also maybe this should not be a warning but a note? I worry that having it as a warning make it look like this is a bad thing to do when actually its expected in the normal workflow that this will be ok.

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 2 above)
Comment (colings86) [reply docs/reference/ilm/policy-definitions.asciidoc]: I think this could be combined with the phases section at the top of the document as it feels weird to have the information split with actions in the middle?

**Review by colings86 — COMMENTED** [review]
I left some comments. I assume documentation explaining how ILM executes policy will be in a separate PR?

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: I will make it a table and slightly reword the warm description

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 5 above)
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: +1

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: +1

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 7 above)
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: +1

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 8 above)
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: I was referring to the fact that they are two distinct actions. replication vs allocation-filtering. I will try and reword

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: ++

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+// NOTCONSOLE
+
+[[ilm-rollover-action]]
+=== Rollover
+
+[WARNING]
+index format must match pattern '^.*-\\d+$', for example (`logs-000001`).
+[WARNING]
+The managed index must set `index.lifecycle.rollover_alias` as the
+alias to rollover. The index must be
```
[the remark this answers is not in the document]
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: ++

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+
+The Rollover Action rolls an alias over to a new index when the existing index
+is considered to be too large or too old.
```
[the remark this answers is not in the document]
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: ++

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+  }
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+==== Example: Rollover when index is too old or too large
+
+This example rolls the index over when it has been created at least
+7 days ago or it is at least 100 gigabytes. In this case, the index will be rolled
+over when at least one of these two conditions is met.
```
[the remark this answers is not in the document]
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: ++

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+  }
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+
+[[ilm-shrink-action]]
+=== Shrink
+
+WARNING: This will make the index read-only (see: <<dynamic-index-settings,index.blocks.write>>)
```
[the remark this answers is not in the document]
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: ++

**Comment by talevy** [comment]
thanks for providing your feedback and adding suggestions to these docs @dakrone and @colings86. updated for another look!

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 15]
```diff
@@ -0,0 +1,505 @@
...
+deleting the index.
+
+`min_age` is usually the time elapsed from the time the index is created. If the
+index is rolled over, then `min_age` is the time elapsed from the time the index is
+rolled over. The intention here is to execute following phases and actions relative to
+when data was written last to a rolled over index.
+
+=== Actions
+
+Here is the list of actions and which phases they are supported by.
```
Comment (colings86) [inline docs/reference/ilm/policy-definitions.asciidoc]: This felt hard to parse for me. Maybe we could reword it to `The below list shows the actions which are available in each phase` ?

File: docs/reference/ilm/policy-definitions.asciidoc [hunk 16]
```diff
@@ -0,0 +1,505 @@
...
+[[ilm-rollover-action]]
+==== Rollover
+
+Phases allowed: hot.
+
+[WARNING]
+index format must match pattern '^.*-\\d+$', for example (`logs-000001`).
+[WARNING]
+The managed index must set `index.lifecycle.rollover_alias` as the
+alias to rollover. The index must also have this alias pointing to it.
```
Comment (colings86) [inline docs/reference/ilm/policy-definitions.asciidoc]: `The index must also have this alias pointing to it` - I think we need to explain more here, because its not enough that the alias points to the index, the index either needs to be the only index int he alias or it needs to be the write index for the alias.

**Review by colings86 — CHANGES_REQUESTED** [review]
I left some comments but could you also make sure all lines are wrapped to 80 characters in this doc?

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 15 above)
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: agreed

File: docs/reference/ilm/policy-definitions.asciidoc (hunk 16 above)
Comment (talevy) [reply docs/reference/ilm/policy-definitions.asciidoc]: good call. I've added a link to the section in the rollover docs talking about this behavior, as well as an example for creating a managed index with an alias and a rollover_alias setting set.

**Review by colings86 — APPROVED** [review]
I left one suggestions but otherwise this LGTM

**Review by dakrone — APPROVED** [review]

**Comment by talevy** [comment]
thanks all for helping with the editing!
